ssl/ssl_sess.c

```diff
@@ -666,7 +663,9 @@ int SSL_CTX_add_session(SSL_CTX *ctx, SSL_SESSION *c)
      */
 
     CRYPTO_THREAD_write_lock(ctx->lock);
-    s = lh_SSL_SESSION_insert(ctx->sessions, c);
+    session_cache = ctx->session_cache;
+    CRYPTO_THREAD_write_lock(session_cache->lock);
```

Comment:
I haven't gone through all the locking changes yet. Just looking for thumbs up/down with respect to the API definition. At which point it will be documented, and the locking will be cleaned up.